### PR TITLE
[WebView] Add support for evaluateJavaScript to WebView

### PR DESCRIPTION
**Pull request by timfpark**
Add support for evaluateJavaScript to the WebView.

### CODE REVIEW

**Comment by timfpark** [comment]
The Travis failure looks unrelated to my commit (testRendersWelcomeScreen) - please let me know if I've misunderstood.

**Comment by spicyj** [comment]
Maybe worth mirroring WKWebView's already-async API here:

https://developer.apple.com/library/ios/documentation/WebKit/Reference/WKWebView_Ref/#//apple_ref/occ/instm/WKWebView/evaluateJavaScript:completionHandler:

**Comment by timfpark** [comment]
Makes sense and definitely like the name better vs. iOS.  So concretely, @spicyj, you are suggesting this signature:

evaluateJavaScript: function(script: string, callback: Function)

that takes a callback of the form:  

function(err, results)

(I flip flopped the err and results vs. WKWebView to match JS conventions)

Let me know and I can make the change--
Tim

**Comment by spicyj** [comment]
Wasn't suggesting any signature in particular, but it would be nice to pick a signature that is flexible and easy to replicate with a WKWebView backend if we ever wanted to switch to that (or make it configurable).

**Comment by timfpark** [comment]
Ok - I'll update the PR with the above.  That seems to balance having a JS fluent interface and the naming of WKWebView.

**Comment by timfpark** [comment]
Ok @spicyj - I update the signature to match up with WKWebView.

**Comment by vjeux** [comment]
@spicyj: on you to review it :)

**Comment by timfpark** [comment]
Any thoughts on this @spicyj?

File: React/Views/RCTWebViewManager.m [hunk 1]
```diff
@@ -79,4 +79,16 @@ - (NSDictionary *)constantsToExport
   }];
 }
 
+RCT_EXPORT_METHOD(evaluateJavaScript:(NSNumber *)reactTag script: (NSString *)script callback:(RCTResponseSenderBlock) callback)
+{
+  [self.bridge.uiManager addUIBlock:^(RCTUIManager *uiManager, RCTSparseArray *viewRegistry) {
+    RCTWebView *view = viewRegistry[reactTag];
+    if (![view isKindOfClass:[RCTWebView class]]) {
+      callback(@[[NSString stringWithFormat:@"Invalid view returned from registry, expecting RKWebView, got: %@", view], [NSNull null]]);
```
Comment (spicyj) [inline React/Views/RCTWebViewManager.m]: This should use `RCTLogError` (which causes a redbox) like the other methods in this file.

**Comment by spicyj** [comment]
Sorry for the delay here.

With the WKWebView API, you can return any (serializable) JS object, not just a string. Can we change this to JSONify the result so that you can return numbers, arrays, and simple objects transparently from a callback?

File: Libraries/Components/WebView/WebView.ios.js [hunk 2]
```diff
@@ -166,6 +166,10 @@ var WebView = React.createClass({
     );
   },
 
+  evaluateJavaScript: function(script: string, callback: Function) {
+    return RCTWebViewManager.evaluateJavaScript(this.getWebWiewHandle(), script, callback);
```
Comment (paramaggarwal) [inline Libraries/Components/WebView/WebView.ios.js]: Typo in `this.getWebWiewHandle()`

File: React/Views/RCTWebViewManager.m (hunk 1 above)
Comment (timfpark) [inline React/Views/RCTWebViewManager.m]: Ok - I will update this to both log and error and callback with the error.

**Comment by timfpark** [comment]
@spicyj:  So what you are thinking is that:

* If the result can be parsed as a number, return it as a number.
* If the result can be parsed as an array, return it as an array.
* If the result can be parsed as a JSON object, return it as an object.
* Otherwise, return it as a string.

Is that right?

File: Libraries/Components/WebView/WebView.ios.js (hunk 2 above)
Comment (timfpark) [inline Libraries/Components/WebView/WebView.ios.js]: The error here is in the definition - it is actually defined getWebWiewHandle in the file.

**Comment by brentvatne** [comment]
reminder ping for @spicyj

**Comment by spicyj** [comment]
If I call `.evaluateJavaScript("foo()", cb)` then I should get whatever `foo` returned back, whether it's a string, a number, a string with numbers in it, or a simple object.

Also, separately: just realized this should probably return a promise.

**Comment by brentvatne** [comment]
Ah yeah, big :+1: to promise!

File: Libraries/Components/WebView/WebView.ios.js (hunk 2 above)
Comment (paramaggarwal) [inline Libraries/Components/WebView/WebView.ios.js]: Why not change in both places then. :)

**Comment by skitterm** [comment]
I would love to see this functionality integrated... Any timetable on the merge?

**Comment by spicyj** [comment]
@skitterm Waiting on @timfpark to make the requested changes.

**Comment by tsunammis** [comment]
+1 with @skitterm

**Comment by spicyj** [comment]
@skitterm @tsunammis If it's important to you, you're more than welcome to submit a pull request of your own.

**Comment by timfpark** [comment]
I'm on it - work has finally let up enough for me to have a go at this.   Give me a few days!

**Comment by timfpark** [comment]
@spicyj:  Ok - I think I incorporated all the feedback:

1. Returns a promise.
2. Parses the string result returned by iOS into its JavaScript type if it can, otherwise returns a string.  
3. Uses RCTLogError to log errors.

**Comment by spicyj** [comment]
No, this doesn't work if my function returns a string like `"2"` – your callback will return number `2` instead.

**Comment by timfpark** [comment]
@spicyj The issue is that stringByEvaluatingJavaScriptFromString (see https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIWebView_Class/) returns a string - so in iOS I don't have the context on if a function that returns "2" is a string or a number - which is why originally I just passed this through as a string in an earlier version of this pull request.   Is there a way that you know about to get this context?

**Comment by spicyj** [comment]
If I ask you to evaluate `foo()`, you can (behind the scenes) pass something like `JSON.stringify(foo())` to `stringByEvaluatingJavaScriptFromString:` and do whatever post-processing you want. For now, JSON serialization is probably a fine choice. This is complicated a bit by the fact that strings like `var x = foo(); x * 2;` should return `4` even though there's no explicit return. This matches the JS semantics for `eval` though, so you can actually run

```
JSON.stringify(eval("var x = foo(); x * 2;"))
```

in your web view and that should pass you back the correct result.

Let me know if that makes any sense.

**Comment by timfpark** [comment]
Great idea @spicyj.  I added that to the PR.

**Comment by spicyj** [comment]
This now doesn't work if the source has a quote mark in it. You should be able to assume that the JSON.parse always succeeds now, right? What if your function returns undefined?

**Comment by paramaggarwal** [comment]
@timfpark @spicyj Correct me if I'm wrong - but what if we simply directly wrap `stringByEvaluatingJavaScriptFromString` only? Looks like this serialisation/deserialisation for possible use in `WKWebView` later is making this brittle.

Later when WKWebView support is added, we'll likely have it side by side, and not replace UIWebView entirely. So maybe it would make sense to have it as a separate component whenever we do decide to support it? What do you guys think?

**Comment by alejomendoza** [comment]
+1

**Comment by odino** [comment]
:+1:

**Comment by resendefreitas** [comment]
+1

**Comment by nevescode** [comment]
+1

**Comment by spicyj** [comment]
I believe that limiting this to strings makes it significantly less useful and it doesn't make sense to match the old iOS API, when the new one is more flexible and can be matched fairly easily by adding a little serialization and deserialization glue on the edges.

**Comment by paramaggarwal** [comment]
My thought was that React's core here could focus on simply exposing the API and then any serialization/deserialization could be done by consumers. Alternatively, in this PR, we could simply hand over the string during input but attempt to infer type and parse the response string on the return value. Currently it is effectively doing an `eval` on an `eval`...

**Comment by stereodenis** [comment]
So, what state of this PR is now?

**Comment by spicyj** [comment]
Not going anywhere.

**Comment by hayeah** [comment]
I've updated @timfpark's changes for v18.0. 

https://github.com/facebook/react-native/compare/master...hayeah:webview-eval-js

Let me see what I can do with the quotation escaping issue...

**Comment by benhoyt** [comment]
This seems like a pretty important feature, especially for hybrid apps. @spicyj Why is this closed and "not going anywhere"?

**Comment by spicyj** [comment]
I laid out some specific concerns/requests and the pull request wasn't updated to address them. That's fine but it means that it didn't get merged. If you submit a new pull request that addresses these concerns then we can look at merging it.

**Comment by benhoyt** [comment]
@spicyj Got it, thanks -- I'd forgotten this was a PR and not a general issue.

**Comment by halilb** [comment]
I'm working on a react-native rich editor component. It's using a simple webview to display and manipulate html and it requires communication between webview and react-native through simple messages. That's why i just need reaching native `evaluateJavaScript` method in Webview.

Here are [iOS](https://github.com/cjwirth/RichEditorView) and [Android](https://github.com/wasabeef/richeditor-android) versions of the component. They're built using the same logic.

@spicyj Please consider my use case for this issue. I'd love to create a pull request for Android and iOS if you agree that we need a simple execute method as a wrapper of native `evaluateJavaScript`.

**Comment by jesperlndk** [comment]
*ping* ... Any updates on this?

**Comment by spicyj** [comment]
Like I said above, this PR is closed because:

> I laid out some specific concerns/requests and the pull request wasn't updated to address them. That's fine but it means that it didn't get merged. If you submit a new pull request that addresses these concerns then we can look at merging it.

No one from the core team is working on this but if a high-quality pull request is submitted (noting the feedback I provided above) then I assure you we'll be happy to consider it.

**Pull request closed without merging**